internal/validator/validator.go

```diff
@@ -0,0 +1,198 @@
+			res = errors.New("Unknown panel kind")
+			break
+		}
+		logrus.Tracef("Matching panel %s against schema: %+v", k, schema)
+
+		// do the validation
+		unified := value.Unify(schema)
+		opts := []cue.Option{
+			cue.Concrete(true),
+			cue.Attributes(true),
+			cue.Definitions(true),
+			cue.Hidden(true),
+		}
+		err = unified.Validate(opts...)
+		if err != nil {
+			logrus.Errorf("%s is not a valid %s panel", k, kind)
+			res = err
+			break
+		}
+	}
+
+	if res == nil {
+		logrus.Info("This dashboard is OK, all its panels are valid")
+	} else {
+		logrus.WithError(res).Error("This dashboard is KO, at least 1 panel is invalid")
+	}
+
+	return res
+}
+
+/*
+ * Load the known list of schemas into the validator
+ */
+func (v *validator) LoadSchemas(path string) {
+	schemas, err := loadSchemas(v.context, v.baseDef, path)
+	if err != nil {
+		logrus.WithError(err).Error("Not able to retrieve the list of schema files")
+		return
+	}
+
+	v.schemas = schemas
+	logrus.Info("Schemas list (re)loaded")
+}
+
+/*
+ * Load & return the known list of schemas
+ */
+func loadSchemas(context *cue.Context, baseDef cue.Value, path string) (map[string]cue.Value, error) {
+	schemas := map[string]cue.Value{}
+
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return schemas, err
+	}
+
+	// process each .cue file to convert it into a CUE Value
+	// for each schema we check that it meets the default specs we expect for any panel, otherwise we dont include it
+	for _, file := range files {
+		schemaPath := fmt.Sprintf("%s/%s", path, file.Name())
+
+		// load the cue files into build.Instances slice
+		buildInstances := load.Instances([]string{}, &load.Config{Dir: schemaPath})
+		// we strongly assume that only 1 buildInstance should be returned (corresponding to the main definition like #panel), otherwise we skip it
+		// TODO can probably be improved
+		if len(buildInstances) != 1 {
+			logrus.Errorf("The number of build instances for %s is != 1, skipping this schema", schemaPath)
+			continue
+		}
+		buildInstance := buildInstances[0]
+
+		// check for errors on the instances (these are typically parsing errors)
+		if buildInstance.Err != nil {
+			logrus.WithError(buildInstance.Err).Errorf("Error retrieving schema for %s, skipping this schema", schemaPath)
+			continue
+		}
+
+		// build Value from the Instance
+		schema := context.BuildInstance(buildInstance)
+		if schema.Err() != nil {
+			logrus.WithError(schema.Err()).Errorf("Error during build for %s, skipping this schema", schemaPath)
+			continue
+		}
+
+		// check if the schema fulfils the base panel requirements
+		unified := baseDef.Unify(schema)
+		if unified.Err() != nil {
+			logrus.WithError(unified.Err()).Errorf("Error during schema validation for %s, skipping this schema", schemaPath)
+			continue
+		}
+
+		// check if another schema for the same Kind was already registered
+		kind, _ := schema.LookupPath(cue.ParsePath("kind")).String()
```

Comment:
what would happen here if `kind` doesn't exist ?

Reply:
in this case the previous if ( check if the schema fulfils the base panel requirements) will trigger an error. The error message is a bit weird in the current state, due to the compromises I had to do for the base panel checks :

_time="2022-04-11T13:16:18+02:00" level=error msg="Error during schema validation for dev/schemas/invalid, skipping this schema" error="field not allowed: kind"_